libs/health-insurance/src/lib/soap/healthInsuranceAPI.ts

```diff
@@ -0,0 +1,222 @@
+      vistaSkjalBody,
+    )}]]>`
+
+    const args = {
+      sendandi: '',
+      tegundskjals: appNumber,
+      skjal: xml,
+    }
+    /*
+      Application statuses:
+      0: annars/hafnað/Rejected
+      1: tókst/Succeeded
+      2: tókst en með athugasemd/Succeeded but with comment
+    */
+    logger.info(`Calling vistaskjal through xroad`)
+    const res: GetVistaSkjalDtoType = await this.xroadCall('vistaskjal', args)
+
+    const vistaSkjal = new VistaSkjalModel()
+    if (!res.VistaSkjalType?.tokst) {
+      vistaSkjal.isSucceeded = false
+      vistaSkjal.comment = res.VistaSkjalType?.villulysing ?? 'Unknown error'
+
+      if (
+        res.VistaSkjalType.villulisti &&
+        res.VistaSkjalType.villulisti.length > 0
+      ) {
+        if (res.VistaSkjalType.villulisti[0].villulysinginnri) {
+          vistaSkjal.comment = res.VistaSkjalType.villulisti[0].villulysinginnri
+        }
+      }
+      logger.error(
+        `Failed to upload document to sjukra because: ${
+          vistaSkjal.comment ?? 'unknown error'
+        }`,
+      )
+
+      return vistaSkjal
+    }
+
+    vistaSkjal.isSucceeded = true
+    vistaSkjal.caseId = res.VistaSkjalType.skjalanumer_si
+    vistaSkjal.comment = res.VistaSkjalType.villulysing ?? ''
+
+    logger.info(`--- Finished applyInsurance api call ---`)
+    return vistaSkjal
+  }
+
+  public async xroadCall(functionName: string, args: object): Promise<any> {
+    // create 'soap' client
+    logger.info(`Start ${functionName} function call.`)
+    const client = await SoapClient.generateClient(
+      this.clientConfig.wsdlUrl,
+      this.clientConfig.baseUrl,
+      this.clientConfig.username,
+      this.clientConfig.password,
+      functionName,
+    )
+    if (!client) {
+      logger.error('HealthInsurance Soap Client not initialized')
+      throw new InternalServerErrorException(
+        'HealthInsurance Soap Client not initialized',
+      )
+    }
+
+    return new Promise((resolve, reject) => {
+      // call 'functionName' function/endpoint
+      client[functionName](args, function (err: any, result: any) {
+        if (err) {
+          logger.error(JSON.stringify(err, null, 2))
+          reject(err)
+        } else {
+          logger.info(`Successful call ${functionName} function`)
+          resolve(result)
+        }
+      })
+    })
+  }
+
+  private OBJtoXML(obj: object) {
+    let xml = ''
+    Object.entries(obj).forEach((entry) => {
+      const [key, value] = entry
+      xml += value instanceof Array ? '' : '<' + key + '>'
+      if (value instanceof Array) {
+        for (const i in value) {
+          xml += '<' + key + '>'
+          xml += this.OBJtoXML(value[i])
+          xml += '</' + key + '>'
+        }
+      } else if (typeof value == 'object') {
+        xml += this.OBJtoXML(new Object(value))
+      } else {
+        xml += value
+      }
+      xml += value instanceof Array ? '' : '</' + key + '>'
+    })
+    return xml
+  }
```

Comment:
This is not a safe xml serializer, eg there's no escaping done.

Is the "skjal" a standalone XML document, i.e. not part of the WSDL?

Reply:
yes, its is. It's a standalone XML